service/controller/resource/prometheusautoscaler/resource.go

```diff
@@ -0,0 +1,106 @@
+package prometheusautoscaler
+
+import (
+	"reflect"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	autoscaling "k8s.io/api/autoscaling/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	vpa_types "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1"
+	vpa_clientset "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/client/clientset/versioned"
+
+	"github.com/giantswarm/prometheus-meta-operator/service/controller/resource/generic"
+	"github.com/giantswarm/prometheus-meta-operator/service/key"
+)
+
+const (
+	Name = "prometheusautoscaler"
+)
+
+type Config struct {
+	VpaClient vpa_clientset.Interface
+	Logger    micrologger.Logger
+}
+
+func New(config Config) (*generic.Resource, error) {
+	clientFunc := func(namespace string) generic.Interface {
+		c := config.VpaClient.AutoscalingV1().VerticalPodAutoscalers(namespace)
+		return wrappedClient{client: c}
+	}
+
+	c := generic.Config{
+		ClientFunc:       clientFunc,
+		Logger:           config.Logger,
+		Name:             Name,
+		GetObjectMeta:    getObjectMeta,
+		GetDesiredObject: getObject,
+		HasChangedFunc:   hasChanged,
+	}
+	r, err := generic.New(c)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	return r, nil
+}
+
+func getObjectMeta(v interface{}) (metav1.ObjectMeta, error) {
+	cluster, err := key.ToCluster(v)
+	if err != nil {
+		return metav1.ObjectMeta{}, microerror.Mask(err)
+	}
+
+	return metav1.ObjectMeta{
+		Name:      "prometheus",
```

Comment:
Should we not name it key.ClusterID(cluster) instead?

Reply:
If you feel strongly about it I will, but it actually bugs me why we name objects after the cluster ID. They're already in a namespace named after the cluster ID so they won't clash, and with the variable name it means I can't re-use commands from history after switching namespaces.

E.g.:

``` sh
# I set current namespace to foobz-prometheus and get the prometheus resource
kubectl ns foobz-prometheus
kubectl get prometheus/foobz

# then I switch to another prometheus namespace and want to get the prometheus resource there
# but I can't simply select the get command from history because the resource has a different name
kubectl ns barbz-prometheus 
kubectl get prometheus/barbz
```

Reply:
TBH, I would prefer if we are consistent. I think calling it prometheus, or prometheus-vpa or prometheus-cr_type would make sense. We could have prometheus-ingress for example. 
The issue here is that if we change that now, I would like if we add an issue in the board to rename resources and maybe document the pattern we want to be consistent

Reply:
I agree about consistency, let's see if others agree about naming and if so I'll add an issue about renaming other things to make them consistent.

I disagree about `prometheus-cr_type` pattern, you already have to specify type when referencing a resource so you have `vpa/prometheus-vpa` or `VerticalPodAutoscaler/prometheus-vpa` which breaks the [golang no stuttering rule](https://blog.golang.org/package-names).